lib/opensearch-aws-sigv4.rb

```diff
@@ -17,6 +17,74 @@
 
 module OpenSearch
   module Aws
+    module Transport
+      # Extending OpenSearch::Transport::Client to sign requests with SigV4.
+      class Sigv4Client < ::OpenSearch::Transport::Client
+        attr_accessor :sigv4_signer
+
+        def initialize(transport_args, sigv4_signer, sigv4_debug, &block)
+          unless sigv4_signer.is_a?(::Aws::Sigv4::Signer)
+            raise ArgumentError, "Please pass a Aws::Sigv4::Signer. A #{sigv4_signer.class} was given."
+          end
+
+          super(transport_args, &block)
+          @sigv4_signer = sigv4_signer
+          @sigv4_debug = sigv4_debug
+          @logger = transport.logger
+        end
+
+        # @see ::OpenSearch::Transport::Client#perform_request
+        def perform_request(method, path, params = {}, body = nil, headers = nil)
+          signature_body = body.is_a?(Hash) ? body.to_json : body.to_s
+          signature = sigv4_signer.sign_request(
+            http_method: method,
+            url: signature_url(path, params),
+            headers: headers,
+            body: signature_body
+          )
+          headers = (headers || {}).merge(signature.headers)
+
+          log_signature_info(signature)
+          super(method, path, params, signature_body, headers)
+        end
+
+        private
+
+        def signature_url(path, params)
+          host = transport.hosts.dig(0, :host)
+          path = "/#{path}" unless path.start_with?('/')
+          query_string = params.empty? ? '' : Faraday::Utils::ParamsHash[params].to_query.to_s
+          URI::HTTP.build(host: host, path: path, query: query_string)
+        end
+
+        # @param [Aws::Sigv4::Signature] signature
+        def log_signature_info(signature)
+          return unless @sigv4_debug
+
+          log('string to sign', signature.string_to_sign)
+          log('canonical request', signature.canonical_request)
+          log('signature headers', signature.headers)
```

Comment:
This class isn't the right place to log request headers, I bet it if you turn on debug logging the transport will do it. I would log the size of data being signed (not the data, it could contain PII!) and the size of the resulting signature. The signature itself will get logged by the transport when sent to the server.